diff --git a/lua/plenary/path.lua b/lua/plenary/path.lua
index a253859..3f61e09 100644
--- a/lua/plenary/path.lua
+++ b/lua/plenary/path.lua
@@ -93,6 +93,15 @@ local function _normalize_path(filename, cwd)
   if filename:match(redundant) then
     filename = filename:gsub(redundant, path.sep)
   end
+  if vim.loop.os_uname().version:find('Windows') and filename:sub(2,2) == ':' and string.find(filename, cwd, 1, true) == 1 then
+    local fn = filename:gsub(cwd, '', 1)
+    if fn:sub(1,1) == '/' or fn:sub(1,1) == '\\' then
+      fn = fn:sub(2)
+    end
+    if fn:sub(2,2) == ':' then
+      filename = fn
+    end
+  end
 
   local out_file = filename
 
@@ -102,6 +111,7 @@ local function _normalize_path(filename, cwd)
     local is_abs = is_absolute(filename, path.sep)
     local split_without_disk_name = function(filename_local)
       local parts = _split_by_separator(filename_local)
+      vim.print(parts)
       -- Remove disk name part on Windows
       if path.sep == "\\" and is_abs then
         table.remove(parts, 1)
